File: ADSA/Assignment 3/VertexCover.java
import java.io.*;
import java.util.*;

public class VertexCover {

    public static void main(String[] args) {
        if (args.length != 2) {
            System.out.println("Usage: java VertexCover <graph_file> <parameter_k>");
            System.exit(1);
        }

        String graphFile = args[0];
        int k = Integer.parseInt(args[1]);
        Graph graph = readGraph(graphFile);
        Result result = vertexCover(graph, k);

        if (result.isCoverFound) {
            System.out.println("Yes");
            System.out.println("Vertex Cover: " + result.cover);
        } else {
            System.out.println("No");
        }
    }

    private static Graph readGraph(String filePath) {
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String[] firstLine = br.readLine().trim().split(" ");
            int n = Integer.parseInt(firstLine[0]);
            int m = Integer.parseInt(firstLine[1]);
            List<int[]> edges = new ArrayList<>();

            String line;
            while ((line = br.readLine()) != null) {
                if (!line.trim().isEmpty()) {
                    String[] parts = line.trim().split(" ");
                    edges.add(new int[]{Integer.parseInt(parts[0]), Integer.parseInt(parts[1])});
                }
            }
            return new Graph( edges);
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    private static Result vertexCover(Graph graph, int k) {
        List<int[]> edges = graph.edges;

        if (k < 0) {
            return new Result(false, new ArrayList<>());
        }

        if (edges.isEmpty()) {
            return new Result(true, new ArrayList<>());
        }

        int[] edge = edges.get(0);
        int u = edge[0];
        int v = edge[1];

        for (int vertex : new int[]{u, v}) {
            List<int[]> newEdges = new ArrayList<>();
            for (int[] e : edges) {
                if (e[0] != vertex && e[1] != vertex) {
                    newEdges.add(e);
                }
            }
            Result result = vertexCover(new Graph(newEdges), k - 1);
            if (result.isCoverFound) {
                result.cover.add(0, vertex);
                return result;
            }
        }

        return new Result(false, new ArrayList<>());
    }

    static class Graph {
        List<int[]> edges;

        Graph(List<int[]> edges) {
            this.edges = edges;
        }
    }

    static class Result {
        boolean isCoverFound;
        List<Integer> cover;

        Result(boolean isCoverFound, List<Integer> cover) {
            this.isCoverFound = isCoverFound;
            this.cover = cover;
        }
    }
}

File: ADSA/Assignment 3/vertex_cover.py
def read_graph(file_path):
    with open(file_path, 'r') as file:
        lines = file.readlines()
    
    # Read number of vertices and edges
    n, m = map(int, lines[0].strip().split())
    edges = []
    
    # Read edges
    for line in lines[1:]:
        if line.strip():  # Ignore blank lines
            u, v = map(int, line.strip().split())
            edges.append((u, v))
    
    return n, edges

def vertex_cover(graph, k):
    n, edges = graph
    if k < 0:
        return False, []

    if not edges:
        return True, []

    # Pick an arbitrary edge
    u, v = edges[0]

    # Create new edge lists for the two cases
    new_edges_with_u = [e for e in edges if u not in e]
    new_edges_with_v = [e for e in edges if v not in e]

    # Explore both possibilities
    cover_with_u = vertex_cover((n, new_edges_with_u), k - 1)
    if cover_with_u[0]:
        return True, [u] + cover_with_u[1]

    cover_with_v = vertex_cover((n, new_edges_with_v), k - 1)
    if cover_with_v[0]:
        return True, [v] + cover_with_v[1]

    # Explore the case where we include both u and v in the cover
    cover_with_both = vertex_cover((n, edges[1:]), k - 2)
    if cover_with_both[0]:
        return True, [u, v] + cover_with_both[1]

    return False, []

def main(file_path, k):
    graph = read_graph(file_path)
    result = vertex_cover(graph, k)

    if result[0]:
        print("Yes")
        print("Vertex Cover:", result[1])
    else:
        print("No")

if __name__ == "__main__":
    import sys
    if len(sys.argv) != 3:
        print("Usage: python vertex_cover.py <graph_file> <parameter_k>")
        sys.exit(1)

    graph_file = sys.argv[1]
    k = int(sys.argv[2])
    main(graph_file, k)